engine/src/main/java/org/terasology/math/SpiralIterable.java

```diff
@@ -0,0 +1,183 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.math;
+
+import com.google.common.base.Preconditions;
+import org.joml.Vector2i;
+import org.joml.Vector2ic;
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * An {@link Iterable} that iterates in
+ * a square-shapes spiral around the central point (inclusive).
+ * <br><br>
+ * The iteration starts in positive x direction.
+ * <br><br>
+ * The iterating vector is reused. <b>Do not attempt to store the instance</b> e.g. in a collection.
+ * @author Martin Steiger
+ */
+public class SpiralIterable implements Iterable<Vector2ic>{
+
+    /**
+     * (MAX_SIDELEN * 2 + 1) ^2 must be < Integer.MAX_VALUE
+     */
+    private static final int MAX_RADIUS = 23169;
+
+    private final Vector2ic center;
+    private final boolean clockwise;
+    private final int maxArea;
+    private final int scale;
+
+    /**
+     * @param center the spiral center
+     * @param clockwise true for clockwise iteration, false for counter-clockwise
+     * @param scale the scale of the iteration (positive integer)
+     * @param maxRadius the maximum radius of the spiral [0..23169] (inclusive)
+     */
+    private SpiralIterable(Vector2ic center, boolean clockwise, int scale, int maxRadius) {
+        Preconditions.checkArgument(scale > 0, "scale must be > 0");
+        Preconditions.checkArgument(maxRadius >= 0, "maxRadius must be >= 0");
+        Preconditions.checkArgument(maxRadius <= MAX_RADIUS, "maxRadius must be <= " + MAX_RADIUS);
+
+        int sideLen = maxRadius * 2 + 1;
+
+        this.scale = scale;
+        this.center = center;
+        this.maxArea = sideLen * sideLen;
+        this.clockwise = clockwise;
+    }
+
+    /**
+     * Iterates in clock-wise orientation around the given point.
+     * The point will be the first iterated point.
+     * @param center the spiral center
+     */
+    public static SpiralIterable.Builder clockwise(Vector2ic center) {
+        return new SpiralIterable.Builder(center, true);
+    }
+
+    /**
+     * Iterates in clock-wise orientation around the given point.
+     * The point will be the first iterated point.
+     * @param center the spiral center
+     */
+    public static SpiralIterable.Builder counterClockwise(Vector2ic center) {
+        return new SpiralIterable.Builder(center, false);
+    }
+
+    @Override
+    public Iterator<Vector2ic> iterator() {
+
+        return new Iterator<Vector2ic>() {
+            private int radius = 1;
+            private int leg;
+            private int x = -1;
+            private int y;
+            private int index;
+
+            private Vector2i pos = new Vector2i();
```

Comment:
I'm not sure about this one...
The client iterates over read only vectors. So he probably doesn't expect their value to change. Each `next()` call however, changes the value of the previously returned vector, since it's always the same instance being returned.
Therefore I think it would be cleaner to return new instance every time. (see next comment)

Reply:
I think the point is to reduce GC by avoiding creating a new instance. @skaldarnar how much does this matter with regards to performance and would it be better to make a new instance?

Reply:
I just realized, that it even points this out in the javadoc: `The iterating vector is reused.`
So it was definitely done consciously. The GC argument is valid (even though we are having tons of unnecessary vector allocations at runtime anyway), so I guess I can live with that.